### ISSUE TITLE
Return whole list instead of empty array if no search term is set

### ISSUE THREAD

**Issue by christianmalek** [krisk/Fuse#229]
I'm using fuse.js with the `includeMatches` option to easily highlight matches. The problem is that the usage of includeMatches changes the structure of the list because it has to includes the matches in a fitting manner:

```js
// with includeMatches (the structure changed to include the matches)
[
  {
    "item": {
      "title": "Old Man's War",
      "author": {
        "firstName": "John",
        "lastName": "Scalzi"
      }
    },
    "matches": [
      {
        "indices": [
          [
            0,
            12
          ]
        ],
        "value": "Old Man's War",
        "key": "title",
        "arrayIndex": 0
      }
    ]
  }
]

// without includeMatches
[
  {
    "title": "Old Man's War",
    "author": {
      "firstName": "John",
      "lastName": "Scalzi"
    }
  }
]
```

My problem is: I want to output the whole list if no search term is set. But fuse.js returns (for logical reasons) an empty array. Instead of this I need the whole list with the matches structure. Otherwise I need to treat it different.

Is it possible to return the whole list with empty matches if no search term is set?

**Comment by krisk** [comment]
This seems like quite an odd request 😅 If not items are found, then the search result **should** be empty. Returning the complete list in the same output structure as if there were any results would be contrary to the purpose of this library.

**Comment by christianmalek** [comment]
You can close this. It's obsolete. :)

**Comment by d3vr** [comment]
This can be useful. Here is an example: suppose you're showing a list of contacts and you want to use a search bar to filter it. If the search term is empty display all the list, otherwise display only the filtered results.

Is there any way to achieve this?

**Comment by devgg** [comment]
I would also be interested in this!

**Comment by ketysek** [comment]
@d3vr definitely mentioned the correct use case

**Comment by ljm42** [comment]
Can we reopen this @krisk ? Getting back the structured array (with "item") when no search term is set would be helpful.

**Comment by Brandoncapecci** [comment]
You can achieve this by just mapping over the input array yourself. The resulting output structure is hardly very different...

```
const correctlyShapedArray = inputArray.map(val => ({
    item: Object.assign(val, {}),
    matches: [],
    score: 1
}));
```

**Comment by chrisrowe** [comment]
+1 on this

**Comment by gjunming** [comment]
Also interested in this.

**Comment by klandell** [comment]
+1

**Comment by RizwanAslam** [comment]
+1

**Comment by acnebs** [comment]
https://github.com/krisk/Fuse/issues/229#issuecomment-441907010 seems to be the best solution for the time being.

As an aside @Brandoncapecci, I'm wondering why you used `item: Object.assign(val, {}),` instead of just `item: val,`?

**Comment by Griefchief** [comment]
+1 as well, this is about to make a mess out of my stateless React component since now I have to send a ton more stuff into it through the stateful wrapper component instead of just setting this as an option in Fuse.

Or to hack it in the stateless react component by accessing Fuse.list array and hope Fuse is meant to be used this way as per Brandoncapecci's comment.

**Comment by fekstr** [comment]
+1

**Comment by reinvanimschoot** [comment]
+1

**Comment by indexcoder** [comment]
+1

**Comment by krisk** [comment]
I wonder if this could be handled simply outside of Fuse. One such way perhaps:

```js
const setup = (items, options) => {
  let fuse = new Fuse(items, options)
  
  const search = patten => {
    let results = fuse.search(pattern)
    if (!results.length) {
      results = items
    }
    return results
  }

  return search
}
```

And you could use it like this:

```js
const search = setup(items, options)

// And search..
let r1 = search(/* query */)
let r2 = search(/* another query */)
```

**Comment by acnebs** [comment]
@krisk I think the use-case here for most people is that they are using Fuse for "filtering" behaviour rather than just searching, in that you have some list of items, the user enters a search term, and now you have a smaller list of items. When it comes to rendering these results to your UI, you don't want to have two different rendering patterns for the different cases – you just want to have a single `for` loop that renders either every item or some subset of the items. This is much easier if Fuse has an option to spit out a "formatted" version of the data when no search term is specified.

**Comment by Winangel** [comment]
I agree with @acnebs, we are using fuse for mention purpose, we need to display results as soon as the user hit @ or # key, the search string is empty at this moment and we need to handle this edge case in a very odd manner when it could just be handled in the same manner as the general case...

**Comment by cshoredaniel** [comment]
I disagree.  Of course I'm biased in that I'm using this for a website search and displaying every page with no search pattern would be exactly the wrong behaviour.  If this is going to done it should be as an option not as the only, or even default, behaviour.

**Comment by acnebs** [comment]
@cshoredaniel I don't think anyone wanted this to be the default behavior (esp. because that would be a breaking change which is unnecessary in this case).

So yes, my own conception of this is that it would be an option, and I assume that applies to everyone else here as well.

**Comment by cshoredaniel** [comment]
@acnebs It wasn't clear from the comments that making it optional was the ask (not explicitly mentioned by anyone, and it looked to me like some were arguing that it was the main use case).  I think as an option there is no problem with this from a user perspective. @krisk might have concerns from maintainability / appropriateness for the library.

**Comment by Tyldar** [comment]
Any chance this is getting attention? We need that feature! :D

**Comment by krisk** [comment]
Thinking about this, I suppose if there’s an option, then yes, we could make it such that when the search string is empty, then the fuse instance would return the entire fuse-formatted list. 

Clearly there’s appetite for this feature. I will thus include this as a minor update for next release.

Now someone please tell me what should that option actually be called? 😅

**Comment by Griefchief** [comment]
returnAllOnEmptyString?

**Comment by tommedema** [comment]
`matchIfEmpty`

**Comment by krisk** [comment]
So, I've been thinking more about this, and have a working implementation. However, I don't think this should be built-in behavior for Fuse, anymore.

Folks have mentioned the use-case of Fuse being used as filtering, as opposed to just search. If so, I think this could be done in userland, by simply applying a custom filtering function directly:

```js
const search = (query) => {
  let results = fuse.search(query)
  // If there are results, great.
  if (results.length) {
    return results
  }

  // If not, create a list of conforming to the Fuse format.
  // (But probably cache it for optimization)
  return myList.map((doc, idx) => ({
    item: doc,
    score: 1,
    refIndex: idx
  }))
}
```

**Comment by acnebs** [comment]
What is the argument against including this behavior in the lib?

Seems like it is a common enough use-case where it makes sense to just include this fairly simple behavior (as your code shows) so that everyone doesn't need to re-implement the same pattern in userland.

**Comment by krisk** [comment]
I mean, I added the feature. You can play around it with it in `fuse.js@6.2.0-beta.1`, by setting `matchEmptyQuery` to `true`.

### My rationale for not wanting this:

**Unexpected behavior**: Is this an appropriate behavior for a search library? I mentioned that I don't think Fuse.js should include things that can be trivially done in userland, while also striving to keep the library purely for searching purposes. For example, would you expect `Array.filter` to return the whole array when no item is matched? Even if it were optional, would that be intuitive? Would you expect the same behavior from other search libraries, or would you create a helper function to enable such?

**Feature creep**: All options and features of Fuse.js are to maximize its fuzzy searching potential. The "return all when empty" option is something that is not necessary for its proper functioning.

**Binary size**: This also contributes to the size of the library, for little value, because, again, it can be done in userland, and the vast majority of use cases don't actually need this. In isolation, this may be totally minor, but we should be careful in setting the precedent that just because something is a common-enough use case, it should be added to the library.

**Performance and maintainability implications**: Generating the formatted list every time isn't ideal, so it would be better to have an internal cached copy. This then requires syncing between copies whenever an item is removed/added. This will start to get complicated and will require re-engineering to accommodate the extra functionality.

### Alternative

Aside from what I suggest in previous comments, maybe what we'd need is more of a plugin system, where people can create their own Fuse.js plugins, and publish them for public use. We could even have a list of official plugins.

(cc: @cshoredaniel for thoughts on this)

**Comment by krisk** [comment]
I mean, I added the feature. You can play around it with it in `fuse.js@6.2.0-beta.1`, by setting `matchEmptyQuery` to `true`.

### My rationale for not wanting this:

**Unexpected behavior**: Is this an appropriate behavior for a search library? I mentioned that I don't think Fuse.js should include things that can be trivially done in userland, while also striving to keep the library purely for searching purposes. For example, would you expect `Array.filter` to return the whole array when no item is matched? Even if it were optional, would that be intuitive? Would you expect the same behavior from other search libraries, or would you create a helper function to enable such?

**Feature creep**: All options and features of Fuse.js are to maximize its fuzzy searching potential. The "return all when empty" option is something that is not necessary for its proper functioning.

**Binary size**: This also contributes to the size of the library, for little value, because, again, it can be done in userland, and the vast majority of use cases don't actually need this. In isolation, this may be totally minor, but we should be careful in setting the precedent that just because something is a common-enough use case, it should be added to the library.

**Performance and maintainability implications**: Generating the formatted list every time isn't ideal, so it would be better to have an internal cached copy. This then requires syncing between copies whenever an item is removed/added. This will start to get complicated and will require re-engineering to accommodate the extra functionality.

### Alternative

Aside from what I suggest in previous comments, maybe what we'd need is more of a plugin system, where people can create their own Fuse.js plugins, and publish them for public use. We could even have a list of official plugins.

### Conclusion

Ultimately, I'll keep this in `beta` release, but it will not be published in `latest`.

(cc: @cshoredaniel for thoughts on this)

**Comment by cshoredaniel** [comment]
@krisk I agree with your rationale for not wanting in Fuse proper.

I don't think a plugin system is quite the right answer though, but
rather a second 'wrapper' library (possibly) with plugins (different
repo):

  * plugins imply even more of a promised API which makes maintenance
etc more difficult.
  * a plugin API tends to want access to (and guarantees from) more of
the internals
  * plugins for enhancing search functionality would fit a search
library; plugins for cosmetics wouldn't IMO.
  * since, so far, most or all of the 'cosmetics'/'extras' can be
implemented in userland they could just as well be implemented in a
wrapper library for those who need a library because of their framework
or whatever.
  * basically the 'fuse-tools' library is a convenience library and
keeping it separate from fuse proper encourages better library design,
keeps the main library smaller, and avoids scope creep in the fuse
library.

Just my 2¢

**Comment by tony** [comment]
I don't want to prolong this thread. I came from flex-seach hoping to find that no search would return all values.

A commenter above says:

> But fuse.js returns (for logical reasons) an empty array.

Search on Amazon or any faceted search, is same as no filter.  In query language, ORMs, absence of a filter returns the contents as if no filter expression exists.

> All options and features of Fuse.js are to maximize its fuzzy searching potential.

I am biased in favor of 1.) Filtering (like other users) but also, 2.) Faceting

When I create an index, the perception to the user is there's already on a search even though there's no term. So having `undefined` | `null` | `""` return the full index is what the listing logic that my implementation - and I believe most others  would expect.

I like to extend fuzziness philosophically to deliver the end user an experience of what they'd expect. The intuition / principle of least astonishment would be - an empty search on a database has infinite fuzziness until something more specific is entered. :smile: 

Example would be going to https://www.amazon.com/b?node=16333372011. No search term lists everything.

**Comment by maxim-c** [comment]
```
{
}
```

**Comment by 9mm** [comment]
I definitely really think there is a good use case for this to be added, for all the above mentioned reasons, and in addition, I have ~45 views in my app, all which cache data. This means that I need to cache TWICE the amount of data, simply because I can't extract the rows out of fuse without applying searching, so I need to store the original data set as well and keep it in memory just so I can show that. This to me is a pretty huge impact for performance, especially when some tables have 15,000 rows. Storing 30K rows simply because I can't retrieve all the data without applying a search feels like pure madness.

I do however thing that most of this thread might be thinking about it from a bit of a backwards way... I wouldn't think of this feature as `return all results when search term is empty`, that would be crazy for a search library and is very mind warping to think about given the context of "searching". I would instead think of it as `show all results if no search term is specified`. Honestly, I would even settle for a method like `getCurrentIndex` or something, and solve it by checking if `search query` is blank or not.

**Comment by james-dillon** [comment]
@krisk +1 for adding a filtering option. (Yes, I'm using fuse for fuzzy filtering). The fact that fuse.js inserts "item" at the top of the tree for every item in the original array makes it quite difficult to manage in React. The restructured data means I need data.id for the original data and data.item.id for the filtered data. From a developer perspective this is very frustrating. 

If you ultimately decide not to make this part of fuse.js, could you please add to the documentation the workaround you devised above? It seems there are many devs who really need this to work, and some chatter online from people trying to make it work. So a note in the docs would help a lot.

**Comment by 9mm** [comment]
yes ^ it also already has half the filtering features built in anyway, such as `findAllMatches`, so feature creep seems weird when you're halfway supporting a feature that isn't even complete

**Comment by cshoredaniel** [comment]
I think this is a perfect opportunity for folks who care about functionality that (mostly) doesn't actually require changes to fuse to set up a repo for an 'fusejs-utils' or 'fusejs-tools' library that implements the functionality outside of fusejs, but gives interested developers a library to use (since that seems to be a wish of many).

Also, I was under the impression that something like getCurrentIndex() already existed in order to enable the reuse of the index.  @krisk Am I mistaken on that?

**Comment by 9mm** [comment]
@cshoredaniel so to be clear, you think a searching/filtering library which can deal with 50,000 rows should need a whole extra library or repo simply to **show the results** without searching? Otherwise, we should store literally twice the amount of data simply to show the results when no search is applied...?

**Comment by cshoredaniel** [comment]
@9mm I may expand this reply later, but for the moment here are a few key points:

1.  From [What is Fuse.js?](https://fusejs.io/#what-is-fuse-js), the very first piece of information in the documentation:
"Fuse.js is a powerful, lightweight fuzzy-search library, with zero dependencies".  There **no** mention of filtering as a library goal.  The fact that many have used it that way simply shows the power of a general solution.
2. As much you appear to be in love with your reinterpretation of the problem, a null query is not the same as an all query and I believe you are conflating the two.
3. I do think it could be argued that it could be beneficial to alter the API so that a 'wrapper' object is used in order to effectively achieve pass by reference instead of pass by value (which is what gives rise to the duplication problem), although I'd ague for backwards compatiblity this should be implemented via a separate entry point.  Also I'd have to recheck the code; I don't recall that the code actually modifies the original array data (which would make pass by reference a problem): @krisk is that a correct recollection?
4. This is not the only feature request which I would anticipate being in an additional library.  That would indeed be absurd.  I believe I have mentioned in previous comments that there are other feature requests that fall outside the scope of the fuse.js library itself, but are of interest.  Those I would anticipate being include in a library such as I have discussed.
5. The 'tone' of your comment, is offensive / a form of ad hominem.  I would suggest you find other ways to express your disagreement.

**Comment by james-dillon** [comment]
@cshoredaniel In reply to your pondering, “I don't recall that the code actually modifies the original array data”: yes, indeed, it does. Fuse returns an array of objects, each of which has a property “item” whose value is a row from the original array. So Fuse basically inserts “item” at the base of the tree. For example, data[0].id in the original would become data[0].item.id after Fuse does its magic. If that were not the case, filtering would be fairly trivial to achieve. 

The obvious reason for this is that Fuse adds additional properties alongside “item,” like a search ranking score. But again it just makes an otherwise trivial filtering usage quite cumbersome.

**Comment by cshoredaniel** [comment]
Sorry it’s been a while since I looked at the code.  To clarify, does the Fuse alter the *original* array or does it create a new one?

It does indeed complicate things if the original array is modified.

Regards,

Daniel

**Comment by cshoredaniel** [comment]
So I took a _brief_ look at the code and from what I could see, without ‘includeAllMatches’ the returned dataset is quite small and is returned via a new array (with the index being used to return a reference to the original data).  Again with just a quick skim it looks like ‘includeAllMatches’ adds data during the search, but again not the whole original ‘object’ (used loosely to mean anything that is not basic type like ‘int’ or ‘string’). If the origin is an array of strings, then the original object is copied because it is a ‘simple’ object.
refIndex AIUI is used to reference the index in the original array for complex objects so that one can retrieve the full original record if one so desires.

Now this was just a skim so I could be misreading, but I don’t think the array is duplicated or modified in Fuse, but that rather if one does not keep one’s own copy of the array it will be garbage collected as it will no longer be referenced.

**Comment by james-dillon** [comment]
@cshoredaniel  Sorry about the confusion. Fuse appears to create a new array with a different object structure from what is seen in the original array. New array (no problem) with new structure (annoying).

**Comment by cshoredaniel** [comment]
Apologies for my slightly over-reactive post to 9mm.  I've updated the post to be more neutral in hopes this doesn't get out of hand.

**Comment by 9mm** [comment]
It's all good Daniel, a lot of times I am unintentionally harsh, and it just doesn't seem that way "in my head at all", but it definitely can from the outside.

**Comment by aleclarson** [comment]
@krisk What about when `minMatchCharLength` equals zero? Could all items be returned then?

**Comment by MegaMech** [comment]
The dev's could consider making the userland solution easier to find (Edit original comment perhaps?). So that those of us going above and beyond to implement fuse in an incorrect manner can find this valuable addition to our pseudo-searches that in-reality are just trivial filters.

Also, plugin/library that disables filtering and only allows fuse to be used for search when?

**Comment by 9mm** [comment]
@MegaMech what is the actual userland solution?

**Comment by MegaMech** [comment]
@9mm This:
```
const searchFilter = (fuse) => {
  const list = myList.map((doc, idx) => ({
    item: doc,
    score: 1,
    refIndex: idx
  }))

  return (query) => {
    let results = fuse.search(query)
    return results.length ? results : list
  }
}
```
I don't like this style though. It's kinda messy to comprehend what's going on.
This is what I did:
```
        if (pattern === "") {
            search_sort_table(d.map((doc, idx) => ({
                item: doc,
                score: 1,
                refIndex: idx
            })));
        } else {
            search_sort_table(fuse.search(pattern));
        }
```

**Comment by Pastromhaug** [comment]
fwiw I also came here expecting to be able to display all results when empty-string is passed in as the search param. Seems like a super common usecase.

**Comment by christopher4lis** [comment]
Is this feature included, or is there another library that has this functionality?

**Comment by celluj34** [comment]
I am also looking for this feature, and am also using it as a filter.

I would posit that it's _easier_ in userland to stop an empty query from returning all, rather than the inverse. Without this feature built-in, I now have to maintain two copies of the data (one for the original data set, and another inside Fuse), and I have to copy/paste more code to patch Fuse to do what I want than simply checking if a string is null/empty, or no code at all if I can provide an option for change this behavior.

**Comment by 9mm** [comment]
@celluj34 yes join the club. I have enormous data sets in memory that I'm dealing with, and require 2 copies simply because a simple function to expose the internal index

**Comment by MegaMech** [comment]
Didn't the devs say if this was supported they would still be making two copies on their side regardless which means there wouldn't be a speed improvement?
At this point I think this issue is popular enough they should just work in a solution of some sort. You could even expose the internal index yourself.

**Comment by celluj34** [comment]
Why would they need two copies?

```
if(search.length === 0)
  return allData;
```

**Comment by 2x2xplz** [comment]
The inability to return all results or to designate a universal wildcard is especially problematic with the Logical Query Operators. Imagine this data set (you can follow along in the [live demo page](https://fusejs.io/demo.html)):

```
[
  { city: "Springfield", state: "MO" },
  { city: "Springfield", state: "KY" },
  { city: "Springfield", state: "MA" },
  { city: "Springfield", state: "NJ" },
  { city: "Springfield", state: "OH" },
  { city: "Toledo", state: "OH" },
  { city: "Cleveland", state: "OH" },
  { city: "Cincinnati", state: "OH" },
  { city: "Columbus", state: "OH" },
]
```

with this setup:

```
const options = {
  threshold: 0.2,
  keys: ["city", "state"]
};
const fuse = new Fuse(list, options);
```

The web page has two search options, City and State. The user types 'SPR' into the City box, and the code converts that into a pattern like:

```
const pattern = { $and: [ { city: "SPR" } , { state: "" } ] }
return fuse.search(pattern) // user wants all cities with SPR, but query returns nothing
```
Since the State search field was empty, the query returns an empty set.

I can already hear you say, obviously, the fix is to use `$or:` instead of `$and:`. Sure, if you only want to allow searching of city OR state. But if the user wants to use both search boxes to get more specific, `$or:` blows up:

```
const pattern = { $or: [ { city: "SPR" } , { state: "OH" } ] }
return fuse.search(pattern) // user wants just the cities with SPR that are in OH, but query returns ALL SPR cities and ALL OH cities.
```
When the user adds more specificity by adding a state, s/he expects the results to be the cities that match both City and State. The `$or:` term goes the opposite way, getting less specific. 

As it stands the only workaround is for the developer to manually construct the search pattern based on whether one or both input fields contain any text. When it's just two inputs, like in this example, that isn't terribly difficult, however, when there are 3, 4 or more search terms (for example, customer id, customer name, customer city -- or player first name, player last name, player position, team name, team city, college) there can be all kinds of permutations.

Making this work properly, along with satisfying the dozens of requests above, simply requires the addition of a `filterMode` option which, when enabled, returns all items (rather than none) when a search field is empty.

I'm going to end this by noting a bit of irony that you're asking aloud how to monetize the library #554, but yet you refuse to add a simple feature that users have been practically *begging* for upwards of 3 years. It shouldn't need to be said, but if you want people to pay to use this library, it should be able to do the things people are asking of it.

**Comment by 9mm** [comment]
@krisk does it still seem like an "odd request"?

**Comment by abdennourzahaf** [comment]
I needed this today, but it seems that it's not added yet.

**Comment by MegaMech** [comment]
This is why my solution should be added to the OP's post. ^

**Comment by acls** [comment]
Using the internal `_docs` is a simple workaround, but possibly frowned upon:
```js
let results;
if (pattern) {
  results = fuse.search(pattern);
} else {
  results = fuse._docs.map((item) => ({item, matches: [], score: 1}));
}
```
So it seems like the easiest solution is to add a `getCollection` function so you don't have to use the internal `_doc`.
... also, a`format` function could be added to make it easier to make the structures match.

**Comment by ryall** [comment]
With TypeScript:

```ts
const searchResults: Fuse.FuseResult<T>[] =
  searchTerm.length
    ? fuse.search(searchTerm)
    : _.map<T, Fuse.FuseResult<T>>(
        items,
        (item, index) => ({
          item,
          refIndex: index,
          matches: [],
          score: 1,
        }),
      );
```

**Comment by Ezra-Siton-UIX** [comment]
I found very very simple solution - by:
` threshold of 1.0 would match anything`
https://fusejs.io/api/options.html#threshold


```
let query = "";
// 1. List of items to search in
const books = [
  {
    title: "Old Man's War",
    author: {
      firstName: 'John',
      lastName: 'Scalzi'
    }
  },
  {
    title: 'The Lock Artist',
    author: {
      firstName: 'Steve',
      lastName: 'Hamilton'
    }
  }
]

let keys = ['title', 'author.firstName'];
// 2. Set up the Fuse instance
const fuse_all_items = new Fuse(books, {
  keys: keys,
  threshold: 1
}) 

const fuse_normal_fuse = new Fuse(books, {
  keys: keys,
  threshold: 0.6 /*deafult value */
}) 

// 3. Now search!
let all_items = fuse_all_items.search('get all items threshold set to 1');
console.log(all_items); /* return [object Arrat] (2) */

let specific_item = fuse_normal_fuse.search('Old Mans War');
console.log(specific_item); /* return [object Arrat] (1) */
```
Also not perfect but also not a mess to create two instances of fuse (One with normal threshold and the other without and use some `if` logic related to search query and so on).

**Comment by nateplusplus** [comment]
We also had to find a hacky solution for this, for anyone who is also having trouble, here's the basic gist of what we did:

If there are no filters or search queries, find an inverse-exact-match of a string that will definitely never appear in our search field.

```js
if (!filters) {
  results = fuse.search({ title: '!1234567890' });
}

```

**Our use case:**

Here is our page which uses Fuse: https://indeed.com/lead/search-results

On this page, we need to show all results if there is no search query or filters applied. Once you search or apply filters, those results will then be filtered appropriately. This has so far been achieved by simply not using fusejs unless there is a search query or filters.

However, we are now adding a new "sort" feature, where a user can click buttons to change the sorting of the results list. We would much rather lean on Fuse to apply that sorting function, rather than use two separate sorting functions. It's unfortunate that the Fuse library does not offer any solution for this fairly common use case.

**Comment by mickaelchanrion** [comment]
> [#229 (comment)](https://github.com/krisk/Fuse/issues/229#issuecomment-441907010) seems to be the best solution for the time being.
> 
> As an aside @Brandoncapecci, I'm wondering why you used `item: Object.assign(val, {}),` instead of just `item: val,`?

I believe this was to create a list of immutable results.

But this way is missing one use case: 
if one of the property of the object is an array or an object, this property is still mutable.

**Comment by vedmant** [comment]
I had the same idea, it would be very useful to have an option to return all results when empty keyword is passed. Currently I need to do if - else, furthermore Fuse returns items wrapped to include matches, which makes me need to transform it manually on search, which is annoying.

**Comment by jmamadeu** [comment]
Well, in my case I did:
`const filtered = !search
    ? data
    : fuse.search(search).map((result) => result.item);`
    
    If the search variable is empty I want to show all of my data by default, if not I want to `fuse` do the search thing!

**Comment by gztomas** [comment]
Without this feature `threshold: 1` does not work as expected as it should "match anything" from the docs. It requires patching from the user space.

**Comment by slinkardbrandon** [comment]
I have another workaround for the time being as well, but I still think the feature(s) I proposed above may be worth implementing, here's a generic implementation of a workaround in typescript that I came up with (anyone can just delete the TS specific syntax/types if they are in raw JS too)

```ts
import Fuse from 'fuse.js';

interface Person {
  firstName: string;
  lastName: string;
}

const data: Person[] = [
  {
    firstName: 'jim',
    lastName: 'halpert'
  },
  {
    firstName: 'michael',
    lastName: 'scott',
  }
];

const fuseIndex = new Fuse(data, {
  includeScore: true,
  includeMatches: true,
  keys: ['firstName', 'lastName'],
});

const search = (term: string): Fuse.FuseResult<Person>[] => {
  if (!term) {
    // Just mimic the return type of `Fuse.FuseResult`
    return data.map((item, refIndex) => ({
      item,
      refIndex,
      matches: [],
      score: 0,
    });
  }

  return fuseIndex.search(term);
}

search('hal') // official response from fuse
search('') // our response with the full data set
```

Hopefully this helps someone out 😄

**Comment by Ikonicity-Airban** [comment]
A simple method I can try is to map through the result OR the original list conditionally, I don't know if achieving this type of functionality makes sense but it works if you don't want to see an empty screen.

```
import Fuse from 'fuse.js';

interface Person {
  firstName: string;
  lastName: string;
}

const data: Person[] = [
  {
    firstName: 'John',
    lastName: 'Smillet'
  },
  {
    firstName: 'Michael',
    lastName: 'Mutt',
  }
];

const fuseIndex = new Fuse(data, {
  includeScore: true,
  includeMatches: true,
  keys: ['firstName', 'lastName'],
});

const search = (term: string): Fuse.FuseResult<Person>[] => {
  if (!term) {
    // Just mimic the return type of `Fuse.FuseResult`
    return data.map((item, refIndex) => ({
      item,
      refIndex,
      matches: [],
      score: 0,
    });
  }

  return fuseIndex.search(term);
}

search('hal') // official response from fuse
search('') // our response with the full data set


function App() => {
}

**Comment by trych** [comment]
Ok, so five years later and literally 100+ people have upvoted this feature. Would be great to have this as an option.

**Comment by HaasEric** [comment]
+1

**Comment by alikleit** [comment]
ping

**Comment by JuSfrei** [comment]
+1

**Comment by arqex** [comment]
This is a bit hacky but it works, if somebody needs inspiration:

```js
const suggestions = inputValue.trim()
      ? fuseIndex.search(inputValue, 5)
      : fuseIndex._docs.slice(0, 5);
```

**Comment by punkpeye** [comment]
> We also had to find a hacky solution for this, for anyone who is also having trouble, here's the basic gist of what we did:
> 
> If there are no filters or search queries, find an [inverse-exact-match](https://fusejs.io/examples.html#extended-search) of a string that will definitely never appear in our search field.
> 
> ```js
> if (!filters) {
>   results = fuse.search({ title: '!1234567890' });
> }
> ```
> 
> **Our use case:**
> 
> Here is our page which uses Fuse: https://indeed.com/lead/search-results
> 
> On this page, we need to show all results if there is no search query or filters applied. Once you search or apply filters, those results will then be filtered appropriately. This has so far been achieved by simply not using Fuse unless there is a search query or filters.
> 
> However, we are now adding a new "sort" feature, where a user can click buttons to change the sorting of the results list. We would much rather lean on Fuse to apply that sorting function, rather than use two separate sorting functions. It's unfortunate that the Fuse library does not offer any solution for this fairly common use case.

This is the best solution in this thread.
